### PR TITLE
Fixed spark_udf to handle models that return pandas.DataFrame.

### PR DESCRIPTION
**Pull request by tomasatdatabricks**

### CODE REVIEW

File: tests/keras/test_keras_model_export.py [hunk 1]
```diff
@@ -85,6 +86,15 @@ def test_model_save_load(model, model_path, data, predicted):
     assert all(pd.read_json(scoring_response.content, orient="records").values.astype(np.float32)
                == predicted)
 
+    # test spark udf
+    spark_udf_preds = score_model_as_udf(os.path.abspath(model_path),
+                                         run_id=None,
+                                         pandas_df=pd.DataFrame(x))
+    np.testing.assert_array_almost_equal(
+        np.array(spark_udf_preds), predicted.reshape(len(spark_udf_preds)), decimal=8)
+
```
Comment (dbczumar) [inline tests/keras/test_keras_model_export.py]: Nit: Spacing (linter will complain)

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -224,6 +224,8 @@ def predict(*args):
         columns = [str(i) for i, _ in enumerate(args)]
         pdf = pandas.DataFrame(schema, columns=columns)
         result = model.predict(pdf)
+        if type(result) == pandas.DataFrame:
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: I think we should use `isinstance(result, pandas.DataFrame)` instead of `type(result) == pandas.DataFrame` here, unless we don't intend to treat subclasses of `pandas.DataFrame` in the same way.

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -224,6 +224,8 @@ def predict(*args):
...
+            return result[result.columns[0]]
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Why do we select the the first column from the output dataframe; do we always expect that this dataframe will only have a single column? If so, we should validate that this is the case and print a useful error message if the dataframe does not conform to our specifications. A comment explaining this restriction would also be useful.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@tomasatdatabricks Glad this is working! Left a few comments

File: mlflow/pyfunc/__init__.py (hunk 3 above)
Comment (tomasatdatabricks) [reply mlflow/pyfunc/__init__.py]: Currently spark_udf can only handle 1-d output. We plan to address that in the future but for now I think returning partial result is better than failing.

File: mlflow/pyfunc/__init__.py (hunk 2 above)
Comment (tomasatdatabricks) [reply mlflow/pyfunc/__init__.py]: Makes sense.

**Comment by dbczumar** [comment]
@tomasatdatabricks Code changes look good. Can we also add docs to the `spark_udf` method and the mlflow.org udf section (https://mlflow.org/docs/latest/models.html#export-a-python-function-model-as-an-apache-spark-udf) about the fact that we only take the first column of the output and why?

File: mlflow/pyfunc/__init__.py [hunk 4]
```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+    from pyspark.sql.types import ArrayType, NumericType, StringType, DataType
+
+    if not isinstance(result_type, DataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+    if not isinstance(elem_type, NumericType) and not isinstance(elem_type, StringType):
+        raise Exception("Invalid requested type ''. pyfunc.spark_udf currently only supports " +
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: We should raise an `MlflowException` here with an appropriate error code from the `mlflow.protos.databricks_pb2` module.

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
         result = model.predict(pdf)
-        return pandas.Series(result)
+        if isinstance(result, pandas.DataFrame):
+            elem_type = result_type
+            if isinstance(result_type, ArrayType):
+                elem_type = result_type.elementType
+            if isinstance(elem_type, NumericType):
+                result = result.select_dtypes(include=("int", "float"))
+                if len(result.columns) == 0:
+                    raise Exception("The requested return type is numeric but the model did not " +
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: We should raise an `MlflowException` here with an appropriate error code from the `mlflow.protos.databricks_pb2` module.

File: mlflow/pyfunc/__init__.py [hunk 6]
```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+
     >>> predict = mlflow.pyfunc.spark_udf(spark, "/my/local/model")
     >>> df.withColumn("prediction", predict("name", "age")).show()
 
     :param spark: A SparkSession object.
     :param path: A path containing a :py:mod:`mlflow.pyfunc` model.
     :param run_id: ID of the run that produced this model. If provided, ``run_id`` is used to
                    retrieve the model logged with MLflow.
+    :param result_type: the return type of the user-defined function. The value can be either a
+                        :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Can we enumerate the classes of acceptable result types and their corresponding UDF behavior? For example:

```
:param result_type: the return type of the user-defined function. The value can be either a
                    :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
                    The following classes of result type are supported:
                           
                        - NumericType: The model must produce numeric output in order to use 
                                       this result type.
                        - Array[NumericType]: If the model produces predictions as a DataFrame, 
                                              ``spark_udf`` will filter the DataFrame to remove 
                                               non-numeric columns before returning.
                        - StringType: ...
                        - Array[StringType]: ...
```

File: mlflow/pyfunc/__init__.py [hunk 7]
```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+
     def predict(*args):
         model = SparkModelCache.get_or_load(archive_path)
         schema = {str(i): arg for i, arg in enumerate(args)}
         # Explicitly pass order of columns to avoid lexicographic ordering (i.e., 10 < 2)
         columns = [str(i) for i, _ in enumerate(args)]
         pdf = pandas.DataFrame(schema, columns=columns)
         result = model.predict(pdf)
-        return pandas.Series(result)
+        if isinstance(result, pandas.DataFrame):
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: We should add tests cases to verify correct handling of Pandas DataFrame outputs for the various acceptable result_types.

File: mlflow/pyfunc/__init__.py [hunk 8]
```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+            elem_type = result_type
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: It looks like lines 253-255 are duplicates of 231-233 and can be removed.

**Review by dbczumar — CHANGES_REQUESTED** [review]

File: mlflow/pyfunc/__init__.py [hunk 9]
```diff
@@ -201,21 +202,65 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
     :return: Spark UDF type returned by the model's prediction method. Default double.
     """
 
     # Scope Spark import to this method so users don't need pyspark to use non-Spark-related
     # functionality.
     from mlflow.pyfunc.spark_model_cache import SparkModelCache
     from pyspark.sql.functions import pandas_udf
+    from pyspark.sql.types import _parse_datatype_string
+    from pyspark.sql.types import DataType, ArrayType, AtomicType
+    from pyspark.sql.types import IntegralType, FractionalType, BooleanType, StringType
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Can we remove `BooleanType`? I don't see it being used anywhere.

File: tests/pyfunc/test_spark.py [hunk 10]
```diff
@@ -54,18 +75,37 @@ def tearDown(self):
...
-        self.assertEqual(178, len(spark_results))
-        for i in range(0, len(pandas_results)):  # noqa
-            self.assertEqual(self._predict[i], pandas_results[i])
-            self.assertEqual(pandas_results[i], spark_results[i]['prediction'])
+        for row in pandas_results.iterrows():
+            assert self._prediction == list(row[1])
+
+        # Test all supported return types
+        type_map = {"double": (DoubleType(), np.number),
+                    "long": (LongType(), np.int),
```
Comment (dbczumar) [inline tests/pyfunc/test_spark.py]: Can we add `float` and `int` here too?

File: mlflow/pyfunc/__init__.py [hunk 11]
```diff
@@ -201,21 +202,65 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+                        - ArrayType(instanceof IntegralType): Return all integer columns. Exception
+                          is raised if there are none.
+                        - "float" or "double" or instanceof FractionalType: The leftmost numeric
+                          result is returned or exception is raised if there is none.
+                        - ArrayType(instanceof FractionalType): Return all numeric columns.
+                          Exception is raised if there are none. No type conversion is performed.
+                        - "string" or StringType: Result is the leftmost column converted to string.
+                        - Array[StringType]: Return all columns converted to string.
+
+                        NOTE: Conversion may fail if e.g. element type is set to IntegerType and the
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Can we specify that the conversion failure comes from PySpark (PyArrow)? How do you feel about using the following text instead?:

```
NOTE: The model output type must be compatible with the specified PySpark ``result_type``. For example, if ``result_type`` is set to IntegerType and the model returns longs instead, PySpark type conversion will fail.
```

File: mlflow/pyfunc/__init__.py [hunk 12]
```diff
@@ -228,7 +273,28 @@ def predict(*args):
         columns = [str(i) for i, _ in enumerate(args)]
         pdf = pandas.DataFrame(schema, columns=columns)
         result = model.predict(pdf)
-        return pandas.Series(result)
+        if not isinstance(result, pandas.DataFrame):
+            result = pandas.DataFrame(data=result)
+
+        if isinstance(elem_type, IntegralType):
+            result = result.select_dtypes(include=np.int)
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: In Python 2, this omits any columns of datatype `np.long`. Can we make this `result.select_dtypes(include=[np.int, np.long])`?

**Review by dbczumar — CHANGES_REQUESTED** [review]
After these changes are addressed, this should be ready to merge!

File: mlflow/pyfunc/__init__.py [hunk 13]
```diff
@@ -201,21 +202,65 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
 
+    The predictions are filtered to contain only the columns that can be represented as the
+    ``result_type``. If the ``result_type`` is string or array of strings, all predictions are
+    converted to string. If the result type is not an array type, the left most column with
+    matching type will be returned.
+
+    Note: Exception is raised if there is no matching type in the result.
+
+    Note: It is recommended to only use one of LongType, DoubleType, StringType or ArrayType of
+    one of these types as the result_type.
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Nit: result_type >> ``result_type``

**Review by dbczumar — APPROVED** [review]
LGTM!

File: mlflow/pyfunc/__init__.py [hunk 14]
```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+
+    supported_types = [IntegerType, LongType, FloatType, DoubleType, StringType]
+
+    if sum([1 if isinstance(elem_type, x) else 0 for x in supported_types]) == 0:
+        raise MlflowException(
+            message="Invalid result_type '{}'. Request type can only be one of or an array of one "
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py]: Nit: Request type >> Result type

File: docs/source/models.rst [hunk 15]
```diff
@@ -499,3 +499,32 @@ Spark cluster and used to score the model.
 
     pyfunc_udf = mlflow.pyfunc.spark_udf(<path-to-model>)
     df = spark_df.withColumn("prediction", pyfunc_udf(<features>))
```
Comment (mateiz) [inline docs/source/models.rst]: Shouldn't we talk about what `result_type` is assumed if you don't pass one? We should say something about that as a transition into the next paragraph. For example, by default, we return the first numeric column as a double or whatever it is.

File: docs/source/models.rst [hunk 16]
```diff
@@ -499,3 +499,32 @@ Spark cluster and used to score the model.
...
+      double is returned or exception is raised if there is no numeric column.
+    * ``ArrayType(FloatType|DoubleType)``: Return all numeric columns cast to the
+      requested. type. Exception is raised if there are numeric columns.
+    * ``'string'`` or ``StringType``: Result is the leftmost column converted to string.
+    * ``Array[StringType]``: Return all columns converted to string.
+
+.. rubric:: Example
+
+.. code:: python
+
```
Comment (mateiz) [inline docs/source/models.rst]: You should show how to import ArrayType and stuff like that from PySpark since it's not clear what types these refer to.

File: docs/source/models.rst [hunk 17]
```diff
@@ -499,3 +499,32 @@ Spark cluster and used to score the model.
 
     pyfunc_udf = mlflow.pyfunc.spark_udf(<path-to-model>)
     df = spark_df.withColumn("prediction", pyfunc_udf(<features>))
+
+The resulting UDF is based Spark's Pandas UDF and is currently limited to producing either a single
+value or an array of values of the same type per observation. You can control what result is
+returned by supplying ``result_type`` argument. Following values are supported:
```
Comment (mateiz) [inline docs/source/models.rst]: "Following values" -> "The following values"

File: docs/source/models.rst [hunk 18]
```diff
@@ -499,3 +499,32 @@ Spark cluster and used to score the model.
...
+    * ``'int'`` or instanceof ``IntegerType``: The leftmost integer that can fit in
```
Comment (mateiz) [inline docs/source/models.rst]: Why would people ever pass an instanceof instead of passing IntegerType directly? It seems easier to just say IntegerType (and same with the other ones below). Also, you may want to link IntegerType to the PySpark page on IntegerType: https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=integertype#pyspark.sql.types.IntegerType

File: mlflow/keras.py [hunk 19]
```diff
@@ -111,9 +111,16 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
...
+
+    from distutils.version import StrictVersion
+
+    if StrictVersion(keras.__version__) >= StrictVersion("2.2.3"):
+        # NOTE: Keras 2.2.3 does not work with unicode paths in python2. Pass in h5py.File instead
+        # of string to avoid issues.
+        with h5py.File(os.path.abspath(model_file), "r") as model_file:
+            return keras.models.load_model(model_file)
+    else:
+        # NOTE: Older versions of Keras only handle filepath.
```
Comment (mateiz) [inline mlflow/keras.py]: Why was this changed? Is it meant to be in this PR? It's OK if this is a bug fix but kind of weird

File: mlflow/pyfunc/__init__.py [hunk 20]
```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
     :param spark: A SparkSession object.
     :param path: A path containing a :py:mod:`mlflow.pyfunc` model.
     :param run_id: ID of the run that produced this model. If provided, ``run_id`` is used to
                    retrieve the model logged with MLflow.
-    :return: Spark UDF type returned by the model's prediction method. Default double.
+    :param result_type: the return type of the user-defined function. The value can be either a
+                        :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
+                        Only a primitive type or an array of primitive types are allowed.
+                        The following classes of result type are supported:
+                        - "int" or instanceof IntegerType: The leftmost integer that can fit in
```
Comment (mateiz) [inline mlflow/pyfunc/__init__.py]: See comment above about instanceof -- do people really subclass IntegerType?

Comment (mateiz) [reply mlflow/pyfunc/__init__.py]: Also it may be good to say `pyspark.sql.types.IntegerType` for this first one so people know what package it's in.

File: mlflow/pyfunc/__init__.py [hunk 21]
```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+                        - ArrayType(IntegerType|LongType): Return all integer columns that can fit
+                          into the requested size.
+                        - "float" or instanceof FloatType: The leftmost numeric result cast to
+                          float32 is returned or exception is raised if there is no numeric column.
+                        - "double" or instanceof DoubleType: The leftmost numeric result cast to
+                          double is returned or exception is raised if there is no numeric column.
+                        - ArrayType(FloatType|DoubleType): Return all numeric columns cast to the
+                          requested. type. Exception is raised if there are numeric columns.
+                        - "string" or StringType: Result is the leftmost column converted to string.
+                        - Array[StringType]: Return all columns converted to string.
```
Comment (mateiz) [inline mlflow/pyfunc/__init__.py]: Shouldn't this be `ArrayType(StringType)`?

File: mlflow/pyfunc/__init__.py [hunk 22]
```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+                        - "long" or instanceof LongType: The leftmost long integer that can fit in
+                           int64 result is returned or exception is raised if there is none.
+                        - ArrayType(IntegerType|LongType): Return all integer columns that can fit
+                          into the requested size.
+                        - "float" or instanceof FloatType: The leftmost numeric result cast to
+                          float32 is returned or exception is raised if there is no numeric column.
+                        - "double" or instanceof DoubleType: The leftmost numeric result cast to
+                          double is returned or exception is raised if there is no numeric column.
+                        - ArrayType(FloatType|DoubleType): Return all numeric columns cast to the
+                          requested. type. Exception is raised if there are numeric columns.
```
Comment (mateiz) [inline mlflow/pyfunc/__init__.py]: typo: `.` after requested

Comment (mateiz) [reply mlflow/pyfunc/__init__.py]: And "exception is raised if there are numeric columns"

File: mlflow/pyfunc/__init__.py [hunk 23]
```diff
@@ -201,20 +203,62 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
...
+    if not isinstance(result_type, DataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+
+    supported_types = [IntegerType, LongType, FloatType, DoubleType, StringType]
+
+    if sum([1 if isinstance(elem_type, x) else 0 for x in supported_types]) == 0:
```
Comment (mateiz) [inline mlflow/pyfunc/__init__.py]: This might be more readable as `if not any([isinstance(elem_type, x) for x in supported_types])`

File: mlflow/keras.py (hunk 19 above)
Comment (tomasatdatabricks) [reply mlflow/keras.py]: Yeah it's a bugfix. The current mlflow version does not work with older Keras versions. I thought it was too small a fix to make into a separate PR but I can extract it out.

Comment (mateiz) [reply mlflow/keras.py]: It's fine as long as it was intended, but it's also okay to have very small PRs.
